### PR TITLE
Unit test with cargo test -- add container and configmap tests

### PR DESCRIPTION
**Pull request by ZichengMa**
Add unit tests to set functions related to container, configmap and some other types.

### CODE REVIEW

File: src/unit_tests/vstd_ext/string_map.rs
Comment (marshtompsxd) [inline src/unit_tests/vstd_ext/string_map.rs:1]: @ZichengMa Can you revert the change in this file? I made a mistake in this test and I have fixed it here https://github.com/vmware-research/verifiable-controllers/pull/279/files

File: src/unit_tests/kubernetes_api_objects/config_map.rs [hunk 1]
```diff
@@ -0,0 +1,37 @@
...
+use crate::pervasive_ext::string_map::*;
+use vstd::prelude::*;
+use vstd::string::*;
+
+verus! {
+
+#[test]
+#[verifier(external)]
+pub fn test_set_metadata() {
+    println!("Testing set_metadata()...");
```
Comment (marshtompsxd) [inline src/unit_tests/kubernetes_api_objects/config_map.rs:15]: Can you remove the println (in this and following tests)?

File: src/unit_tests/kubernetes_api_objects/config_map.rs [hunk 2]
```diff
@@ -0,0 +1,37 @@
...
+    let object_meta_gen = || {
```
Comment (marshtompsxd) [inline src/unit_tests/kubernetes_api_objects/config_map.rs:16]: I don't think you need this closure. Just use object_meta.clone() -- if it does not implement clone then you can implement one for it. See https://github.com/vmware-research/verifiable-controllers/blob/main/src/kubernetes_api_objects/config_map.rs#L43

File: src/unit_tests/kubernetes_api_objects/container.rs [hunk 3]
```diff
@@ -0,0 +1,357 @@
...
+    exec_action.set_command(vec![new_strlit("command").to_string()]);
+    assert_eq!(vec!["command".to_string()], exec_action.into_kube().command.unwrap());
+}
+
+#[test]
+#[verifier(external)]
+pub fn test_set_exec_lifecyclehandler() {
+    println!("Testing set_exec() for lifecycle handler...");
+    let mut handler = LifecycleHandler::default();
+    let exec_action_gen = ||{
```
Comment (marshtompsxd) [inline src/unit_tests/kubernetes_api_objects/container.rs:138]: You can implement a clone for ExecAction instead of using the closure here

File: src/unit_tests/kubernetes_api_objects/container.rs [hunk 4]
```diff
@@ -0,0 +1,357 @@
...
+    };
+    container.set_ports(ports());
+    assert_eq!(ports().into_iter().map(|v: ContainerPort| v.into_kube()).collect::<Vec<_>>(),
+               container.into_kube().ports.unwrap());
+}
+
+// Tests for life cycle
+#[test]
+#[verifier(external)]
+pub fn test_set_command_execaction() {
```
Comment (marshtompsxd) [inline src/unit_tests/kubernetes_api_objects/container.rs:126]: Just call it set_command

File: src/unit_tests/kubernetes_api_objects/container.rs [hunk 5]
```diff
@@ -0,0 +1,357 @@
...
+    assert_eq!(exec_action_gen().into_kube(), handler.into_kube().exec.unwrap());
+}
+
+#[test]
+#[verifier(external)]
+pub fn test_set_pre_stop() {
+    println!("Testing set_pre_stop()...");
+    let mut lifecycle = Lifecycle::default();
+    let handler_gen = ||{
+        let mut handler = LifecycleHandler::default();
```
Comment (marshtompsxd) [inline src/unit_tests/kubernetes_api_objects/container.rs:153]: You can implement a clone for LifecycleHandler

File: src/unit_tests/kubernetes_api_objects/container.rs [hunk 6]
```diff
@@ -0,0 +1,357 @@
...
+    lifecycle.set_pre_stop(handler_gen());
+    assert_eq!(handler_gen().into_kube(), lifecycle.into_kube().pre_stop.unwrap());
+}
+
+#[test]
+#[verifier(external)]
+pub fn test_set_lifecycle() {
+    println!("Testing set_lifecycle()...");
+    let mut container = Container::default();
+    let lifecycle_gen = ||{
```
Comment (marshtompsxd) [inline src/unit_tests/kubernetes_api_objects/container.rs:168]: You can implement a clone for Lifecycle

File: src/unit_tests/kubernetes_api_objects/container.rs [hunk 7]
```diff
@@ -0,0 +1,357 @@
...
+    assert_eq!(lifecycle_gen().into_kube(), container.into_kube().lifecycle.unwrap());
+}
+
+// Tests for resources
+#[test]
+#[verifier(external)]
+pub fn test_set_resources() {
+    println!("Testing set_resources()...");
+    let mut container = Container::default();
+    let resources_gen = ||{
```
Comment (marshtompsxd) [inline src/unit_tests/kubernetes_api_objects/container.rs:187]: Implement a clone for ResourceRequirements

File: src/unit_tests/kubernetes_api_objects/container.rs [hunk 8]
```diff
@@ -0,0 +1,357 @@
...
+    assert_eq!(deps_hack::k8s_openapi::apimachinery::pkg::util::intstr::IntOrString::Int(8080),
+               tcp_socket_action.into_kube().port);
+}
+
+#[test]
+#[verifier(external)]
+pub fn test_set_exec_probe(){
+    println!("Testing set_exec() for probe...");
+    let mut probe = Probe::default();
+    let exec_action_gen = ||{
```
Comment (marshtompsxd) [inline src/unit_tests/kubernetes_api_objects/container.rs:223]: Implement clone for ExecAction

File: src/unit_tests/kubernetes_api_objects/container.rs [hunk 9]
```diff
@@ -0,0 +1,357 @@
...
+    probe.set_success_threshold(3);
+    assert_eq!(3, probe.into_kube().success_threshold.unwrap());
+}
+
+#[test]
+#[verifier(external)]
+pub fn test_set_tcp_socket() {
+    println!("Testing set_tcp_socket()...");
+    let mut probe = Probe::default();
+    let tcp_socket_action_gen = ||{
```
Comment (marshtompsxd) [inline src/unit_tests/kubernetes_api_objects/container.rs:273]: Implement a clone for TCPSocketAction

File: src/unit_tests/kubernetes_api_objects/container.rs [hunk 10]
```diff
@@ -0,0 +1,357 @@
...
+    container.set_liveness_probe(probe_gen());
+    assert_eq!(probe_gen().into_kube(), container.into_kube().liveness_probe.unwrap());
+}
+
+#[test]
+#[verifier(external)]
+pub fn test_set_readiness_probe() {
+    println!("Testing set_readiness_probe()...");
+    let mut container = Container::default();
+    let probe_gen = ||{
```
Comment (marshtompsxd) [inline src/unit_tests/kubernetes_api_objects/container.rs:314]: Implement clone for Probe

File: src/unit_tests/kubernetes_api_objects/container.rs [hunk 11]
```diff
@@ -0,0 +1,357 @@
...
+    println!("Testing set_image_pull_policy()...");
+    let mut container = Container::default();
+    container.set_image_pull_policy(new_strlit("image_pull_policy").to_string());
+    assert_eq!("image_pull_policy".to_string(), container.into_kube().image_pull_policy.unwrap());
+}
+
+// Test for Envs
+#[test]
+#[verifier(external)]
+pub fn test_set_env_() {
```
Comment (marshtompsxd) [inline src/unit_tests/kubernetes_api_objects/container.rs:348]: Why is there a suffix `_` in the test name?

File: src/unit_tests/kubernetes_api_objects/container.rs
Comment (marshtompsxd) [inline src/unit_tests/kubernetes_api_objects/container.rs:1]: @ZichengMa This file has tests for different structs, like Container and Probe and so on. Can you divide them into different modules and put the tests in corresponding modules? Name the module as `Container`, `Probe` and so on

File: src/unit_tests/kubernetes_api_objects/container.rs (hunk 4 above)
Comment (ZichengMa) [reply src/unit_tests/kubernetes_api_objects/container.rs:126]: There are two set_command need to be tested, so I add a tail for this one. The other one is set_command for container.

Comment (marshtompsxd) [reply src/unit_tests/kubernetes_api_objects/container.rs:126]: Then you can use different mod to separate tests for different structs

**Comment by marshtompsxd** [comment]
@ZichengMa Pls resolve the conflicts also

File: src/unit_tests/kubernetes_api_objects/container.rs (hunk 4 above)
Comment (ZichengMa) [reply src/unit_tests/kubernetes_api_objects/container.rs:126]: I just tried to keep the same structure as our kubernetes_api_objects directory. Wouldn't keeping the same layer/structure make the tests easier to be found/read?

Comment (marshtompsxd) [reply src/unit_tests/kubernetes_api_objects/container.rs:126]: No I am not asking you to divide them into different files. Keep them in the same file, but different mod. See http://web.mit.edu/rust-lang_v1.25/arch/amd64_ubuntu1404/share/doc/rust/html/book/second-edition/ch07-01-mod-and-the-filesystem.html#module-definitions -- you will have one module for one struct you are testing. Also see my comment before https://github.com/vmware-research/verifiable-controllers/pull/278#discussion_r1328972589

Comment (ZichengMa) [reply src/unit_tests/kubernetes_api_objects/container.rs:126]: Sure, I will fix it.

**Comment by marshtompsxd** [comment]
@ZichengMa You might need to rebase main again

**Comment by marshtompsxd** [comment]
@ZichengMa Did you divide the container.rs into different modules?

**Comment by ZichengMa** [comment]
> @ZichengMa Did you divide the container.rs into different modules?

Yes

**Comment by marshtompsxd** [comment]
> > @ZichengMa Did you divide the container.rs into different modules?
> 
> Yes

You divide them into different files?

**Comment by marshtompsxd** [comment]
Anyway pls rebase and push again. I cannot merge it before all checks pass

**Comment by marshtompsxd** [comment]
move to #294
